File: ruby/directory_scanner.rb
require 'time'

class DirectoryScanner
  OUTPUT_FILE_PATH = '/home/usuario/outfiles/output_ruby.txt'
  EXECUTION_TIMES_FILE = '/home/usuario/execution_times.txt'

  def self.main
    start_time = Time.now
    repeated = '*' * 70
    directory_path = '/home'
    file_paths = []


    read_files_recursively(directory_path, file_paths)


    write_files(file_paths)

    end_time = Time.now


    duration = end_time - start_time
    hours = (duration / 3600).to_i
    minutes = ((duration % 3600) / 60).to_i
    seconds = (duration % 60).to_i
    formatted_duration = format('%02d:%02d:%02d', hours, minutes, seconds)


    total_memory = `free -b | grep Mem | awk '{print $2}'`.to_i
    free_memory = `free -b | grep Mem | awk '{print $4}'`.to_i
    used_memory = total_memory - free_memory
    process_cpu_time = `ps -p #{Process.pid} -o cputime=`.strip


    File.open(EXECUTION_TIMES_FILE, 'a') do |writer|
      writer.puts
      writer.puts repeated
      writer.puts 'Ruby'
      writer.puts "Start: #{start_time.strftime('%H:%M:%S')}"
      writer.puts "End: #{end_time.strftime('%H:%M:%S')}"
      writer.puts "Execution time: #{formatted_duration}"
      writer.puts "Memory usage (in bytes): #{used_memory}"
      writer.puts "CPU time in user mode: #{process_cpu_time}"
      writer.puts "File with listed files: #{OUTPUT_FILE_PATH}"
      writer.puts repeated
    end

    puts "Add line in time file: #{EXECUTION_TIMES_FILE}"
  end

  def self.write_files(file_paths)
    File.open(OUTPUT_FILE_PATH, 'w') do |writer|
      file_paths.each do |path|
        writer.puts path
      end
    end
  rescue StandardError => e
    puts "Error when trying to write: #{e.message}"
  end

  def self.read_files_recursively(directory, file_paths)
    Dir.foreach(directory) do |file|
      next if file == '.' || file == '..'

      file_path = File.join(directory, file)
      if File.directory?(file_path)
        read_files_recursively(file_path, file_paths)
      else
        file_paths << file_path
      end
    end
  rescue StandardError => e
    puts "Error when reading directory: #{e.message}"
  end
end


DirectoryScanner.main

File: java/src/main/java/DirectoryScanner.java
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.time.Duration;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.lang.management.ManagementFactory;
import com.sun.management.OperatingSystemMXBean;


public class DirectoryScanner {

    public static String outputFilePath = "/home/usuario/outfiles/output_java.txt";

    public static void main(String[] args)   {
        LocalTime startTime = LocalTime.now();
        OperatingSystemMXBean osBean = (OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean();
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("HH:mm:ss");
        String repeated = "*".repeat(70);
        String directoryPath = "/home";
        List<String> filePaths = new ArrayList<>();
        readFilesRecursively(new File(directoryPath), filePaths);
        writeFiles(filePaths);
        LocalTime endTime = LocalTime.now();

        String filePathExecute="/home/usuario/execution_times.txt";

        Duration duration = Duration.between(startTime, endTime);
        long hours = duration.toHours();
        long minutes = duration.toMinutes() % 60;
        long seconds = duration.getSeconds() % 60;
        String formattedDuration = String.format("%02d:%02d:%02d", hours, minutes, seconds);

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filePathExecute, true))) {
            double processCpuLoad = osBean.getProcessCpuLoad() * 100;
            long processCpuTime = osBean.getProcessCpuTime();
            long totalMemorySize = osBean.getTotalPhysicalMemorySize();
            long freeMemorySize = osBean.getFreePhysicalMemorySize();
            long usedMemorySize = totalMemorySize - freeMemorySize;
            writer.newLine();
            writer.write(repeated);
            writer.newLine();
            writer.write("Java");
            writer.newLine();
            writer.write(String.format("Start: %s", startTime.format(formatter)));
            writer.newLine();
            writer.write(String.format("End: %s", endTime.format(formatter)));
            writer.newLine();
            writer.write(String.format("Execution time: %s", formattedDuration));
            writer.newLine();
            writer.write(String.format("Memory usage (in bytes): %s",usedMemorySize ));
            writer.newLine();
            writer.write(String.format("CPU time in user mode: %s",processCpuTime));
            writer.newLine();
            writer.write(String.format("Kernel Mode CPU Time: %s",freeMemorySize));
            writer.newLine();
            writer.write(String.format("File with listed files: %s",outputFilePath));
            writer.newLine();
            writer.write(repeated);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }


        System.out.println("Add line in time file: " + filePathExecute);

    }

    private static void writeFiles(List<String> filePaths) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputFilePath))) {

            for (String path : filePaths) {
                writer.write(path);
                writer.newLine();
            }

        } catch (IOException e) {
            System.err.println("Error when try write: " + e.getMessage());
        }

    }

    private static void readFilesRecursively(File directory, List<String> filePaths) {
        if (directory.isDirectory()) {
            File[] files = directory.listFiles();
            if (files != null) {
                for (File file : files) {
                    if (file.isDirectory()) {
                        readFilesRecursively(file, filePaths);
                    } else {
                        filePaths.add(file.getAbsolutePath());
                    }
                }
            }
        }
    }
}
